### PR TITLE
Added ability to limit constraints by device

### PR DESCRIPTION
**Pull request by ghost**
Similar to setting the offset (.with.offset(...)), a user can now also use .with.device(...) to limit what device constraints are applied to.

New constants have been introduced to support this (MASDevice....).

i.e.
make.left.equalTo(self.view.mas_left).with.device(MASDeviceiPhone6);
make.left.equalTo(self.view.mas_left).with.offset(30.0f).with.device(MASDeviceiPhone6Plus)

### CODE REVIEW

**Comment by robertjpayne** [comment]
@igorefremov interesting idea, I actually do stuff like this from time to time though normally more related to sizes than devices. @cloudkite whats your thoughts?

**Comment by igorefremov** [comment]
@robertjpayne This actually uses the screen size to determine the device.  The reason I created individual device constants was to make it easier to recognize what constraints will be be applied and when.  With that being said, I will be pushing a commit shortly removing many of the iPad constants because most of them share the same screen size, and there's no way to determine individual iPad models that share the same screen size.

**Comment by cloudkite** [comment]
@igorefremov @robertjpayne Personally would prefer to mimic the IB concept of size classes.

image

While a bit more abstract than specifying an actual device it lends itself better to future proofing and responsive design.

What are your thoughts?

**Comment by cloudkite** [comment]
some good references

https://developer.apple.com/library/ios/documentation/UserExperience/Conceptual/MobileHIG/LayoutandAppearance.html
http://useyourloaf.com/blog/size-classes.html

**Comment by robertjpayne** [comment]
@cloudkite I find size classes totally worthless personally. All phones in portrait are considered 'Compact' but hardly can you use the same layout on a 6S+ and a 4S.

To be honest I've been thinking about more of a CSS media query breakpoints API based strictly on widths, it's far easier to reason with and more flexible than the size classes.

**Comment by cloudkite** [comment]
@robertjpayne fair enough, I haven't used size classes myself but sounded good in theory!

media breakpoints API sounds good

**Comment by robertjpayne** [comment]
@cloudkite yea I mean really size classes are just media breakpoints except you only get two! They are good for deciding when to do split screen vs not is all I've found ;)

**Comment by robertjpayne** [comment]
@igorefremov not to Hijack your thread, but I want to do some more thinking about this. I think there is maybe an API to be had for conditional constraints such as:

`make.left.equalTo(otherView.mas_right).when(<block evaluator OR NSNumber bool>);`

This would be a bit more flexible, and someone could create constants for different conditions like:

`#define MAS_WHEN_IPHONE6 @([UIScreen mainScreen].bounds.size.width == 750)`

Or

`BOOL (^iPhone6)(void) = ^BOOL { return [UIScreen mainScreen].bounds.size.width == 750 };`

Ideally we'd run block evaluators whenever a view's `updateConstraints` method is called, and if it returns a different value we'd activate/deactivate the constraint. This would have a huge benefit to avoiding a lot of `mas_remakeConstraints` API's just to reflow due to orientation changes or keyboard etc…

**Comment by robertjpayne** [comment]
Alternatively the API may be cleaner if the `when` is on the left hand side of the make:

```
make.when(@YES)(^(MASConstraintMaker *make) {
  make.left.equalTo(50);
});
```

**Comment by igorefremov** [comment]
@robertjpayne @cloudkite I agree with both of you.  I definitely like the .when(<block>) syntax.  In my fork, I plan on keeping the .with.device syntax and implementing the .when syntax at some point in the near future.  The reason I like the .with.device syntax so much is because it helped us greatly with our current app, which needed slightly different margins on iPhone 5's and iPhone 6's, while also maintaining the readability that Masonry offers.  95% of the time that I use Masonry, there is no "decision" code inside of the make blocks.  It is simple make.left...., make.right...., etc, and that's it.

**TL;DR** The .with.device syntax is simply clean and quick, but the .when(<block>) syntax is definitely something I could see implementing and using for our current project.

**Comment by robertjpayne** [comment]
Closing this as I'm not willing to accept it, mostly it's a huge additional feature at a time when we're trying to keep Masonry pretty minimal and stick with what API's are already available.
